File: backend/src/modules/misiones/misiones.controller.js
// src/modules/misiones/misiones.controller.js
// import misionesService from './misiones.service.js';

const misionesController = {};
const JAVA_API = 'http://localhost:8080/api/java/misiones';

misionesController.obtenerMisiones = async (req, res) => {
    try {
        const response = await fetch(JAVA_API);
        const data = await response.json();
        return res.status(200).json(data);
    } catch (error) {
        return res.status(500).json({ error: 'Java backend no disponible' });
    }
};

misionesController.obtenerEjerciciosPorMision = async (req, res) => {
    const { id } = req.params;

    try {
        const response = await fetch(`${JAVA_API}/${id}/ejercicios`);
        const data = await response.json();
        if (!response.ok) return res.status(response.status).json(data);
        return res.status(200).json(data);
    } catch (error) {
        return res.status(500).json({ error: 'Java backend no disponible' });
    }
};

misionesController.crearMision = async (req, res) => {
    const { nombre } = req.body;

    if (!nombre) {
        return res.status(400).json({ msg: 'Se requiere el nombre de la misión' });
    }

    try {
        const response = await fetch(JAVA_API, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ nombre })
        });
        const data = await response.json();
        if (!response.ok) return res.status(response.status).json(data);
        return res.status(201).json(data);
    } catch (error) {
        return res.status(500).json({ error: 'Java backend no disponible' });
    }
};

misionesController.crearEjercicio = async (req, res) => {
    const { id } = req.params;
    const { pregunta, opA, opB, opC, opD, correcta } = req.body;

    if (!pregunta || !opA || !opB || !opC || !opD || !correcta) {
        return res.status(400).json({ msg: 'Faltan campos obligatorios para crear el ejercicio' });
    }

    try {
        const response = await fetch(`${JAVA_API}/${id}/ejercicios`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(req.body)
        });
        const data = await response.json();
        if (!response.ok) return res.status(response.status).json(data);
        return res.status(201).json(data);
    } catch (error) {
        return res.status(500).json({ error: 'Java backend no disponible' });
    }
};

export default misionesController;

File: java-backend/src/main/java/com/example/mathexpedition/controllers/MisionesController.java
package com.example.mathexpedition.controllers;

import com.example.mathexpedition.models.Ejercicio;
import com.example.mathexpedition.models.Mision;
import com.example.mathexpedition.repository.EjercicioRepository;
import com.example.mathexpedition.repository.MisionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/misiones")
@CrossOrigin(origins = "*")
public class MisionesController {

    @Autowired
    private MisionRepository repoMisiones;

    @Autowired
    private EjercicioRepository repoEjercicios;

    @GetMapping
    public ResponseEntity<List<Mision>> obtenerMisiones() {
        return ResponseEntity.ok(repoMisiones.obtenerTodos());
    }

    @GetMapping("/{id}/ejercicios")
    public ResponseEntity<Object> obtenerEjerciciosPorMision(@PathVariable String id) {
        Mision mision = repoMisiones.obtenerPorId(id);
        if (mision == null) {
            Map<String, String> error = new HashMap<>();
            error.put("msg", "No se encontró la misión con ID " + id);
            return ResponseEntity.status(404).body(error);
        }

        List<Ejercicio> ejercicios = repoEjercicios.obtenerTodos();
        List<Ejercicio> ejerciciosDeLaMision = ejercicios.stream()
                .filter(e -> mision.getEjerciciosIds() != null && mision.getEjerciciosIds().contains(e.getId()))
                .toList();

        return ResponseEntity.ok(ejerciciosDeLaMision);
    }

    @PostMapping
    public ResponseEntity<Map<String, Object>> crearMision(@RequestBody Map<String, Object> body) {
        String nombre = (String) body.get("nombre");
        if (nombre == null) {
            Map<String, Object> error = new HashMap<>();
            error.put("msg", "Se requiere el nombre de la misión");
            return ResponseEntity.badRequest().body(error);
        }

        List<Mision> misiones = repoMisiones.obtenerTodos();
        for (Mision m : misiones) {
            if (m.getNombre() != null && m.getNombre().equalsIgnoreCase(nombre)) {
                Map<String, Object> error = new HashMap<>();
                error.put("msg", "Ya existe una misión con ese nombre");
                return ResponseEntity.badRequest().body(error);
            }
        }

        long nuevoIdLong = misiones.isEmpty() ? 1 : misiones.stream().mapToLong(m -> {
            try { return Long.parseLong(m.getIdMision()); } catch (Exception e) { return 0; }
        }).max().orElse(0) + 1;
        String nuevoId = String.valueOf(nuevoIdLong);

        Mision nuevaMision = new Mision(nuevoId, nombre, "", null);
        repoMisiones.agregarMision(nuevaMision);

        Map<String, Object> success = new HashMap<>();
        success.put("msg", "Misión creada exitosamente");
        success.put("mision", nuevaMision);
        return ResponseEntity.status(201).body(success);
    }

    @PostMapping("/{id}/ejercicios")
    public ResponseEntity<Map<String, Object>> crearEjercicio(@PathVariable String id, @RequestBody Map<String, Object> body) {
        String pregunta = (String) body.get("pregunta");
        String opA = (String) body.get("opA");
        String opB = (String) body.get("opB");
        String opC = (String) body.get("opC");
        String opD = (String) body.get("opD");
        String correcta = (String) body.get("correcta");

        if (pregunta == null || opA == null || opB == null || opC == null || opD == null || correcta == null) {
            Map<String, Object> error = new HashMap<>();
            error.put("msg", "Faltan campos obligatorios para crear el ejercicio");
            return ResponseEntity.badRequest().body(error);
        }

        Mision mision = repoMisiones.obtenerPorId(id);
        if (mision == null) {
            Map<String, Object> error = new HashMap<>();
            error.put("msg", "La misión destino no existe");
            return ResponseEntity.status(404).body(error);
        }

        List<Ejercicio> ejercicios = repoEjercicios.obtenerTodos();
        long nuevoIdLong = ejercicios.isEmpty() ? 1 : ejercicios.stream().mapToLong(e -> {
            try { return Long.parseLong(e.getId()); } catch (Exception ex) { return 0; }
        }).max().orElse(0) + 1;
        String nuevoId = String.valueOf(nuevoIdLong);

        Ejercicio nuevoEjercicio = new Ejercicio(nuevoId, "1", pregunta, opA, opB, opC, opD, correcta, id);
        repoEjercicios.agregarEjercicio(nuevoEjercicio);

        mision.agregarEjercicio(nuevoId);
        repoMisiones.actualizarMision(id, mision);

        Map<String, Object> success = new HashMap<>();
        success.put("msg", "¡Ejercicio creado e inyectado a la misión con éxito!");
        success.put("ejercicio", nuevoEjercicio);
        return ResponseEntity.status(201).body(success);
    }
}
